File: leetcode75/go/643-maximum-average-subarray-1.go
package main

import (
	"fmt"
	"math"
)

func main() {
	nums := []int{1, 12, -5, -6, 50, 3}
	k := 4
	fmt.Println(findMaxAverage(nums, k))

	nums2 := []int{5}
	k2 := 1
	fmt.Println(findMaxAverage(nums2, k2))
}

func findMaxAverage(nums []int, k int) float64 {
	maxAverage := float64(0)
	windowSum := 0

	for i := 0; i < k; i++ {
		windowSum += nums[i]
	}

	maxAverage = float64(windowSum) / float64(k)

	for i := k; i < len(nums); i++ {
		windowSum += nums[i] - nums[i-k]
		currentAverage := float64(windowSum) / float64(k)
		maxAverage = math.Max(maxAverage, currentAverage)
	}

	return maxAverage
}

File: leetcode75/js/643-maximum-average-subarray-1.js
var findMaxAverage = function(nums, k) {
    let sum = 0
    for (let i = 0; i < k; i++) {
        sum += nums[i]
    }
    let maxAverage = sum / k
    for (let i = k; i < nums.length; i++) {
        sum = sum - nums[i - k] + nums[i]
        maxAverage = Math.max(maxAverage, sum / k)
    }
    return maxAverage    
}

var nums = [1,12,-5,-6,50,3], k = 4
console.log(findMaxAverage(nums, k)) // 12.75
var nums = [5], k = 1
console.log(findMaxAverage(nums, k)) // 5
var nums = [0,4,0,3,2], k = 1
console.log(findMaxAverage(nums, k)) // 4